pkg/consensus/consensus.go

```diff
@@ -337,11 +337,16 @@ events:
 		case b := <-s.blockEvents:
 			s.handleChainBlock(b)
 		}
-		// Always process block event if there is any, we can add one above.
-		select {
-		case b := <-s.blockEvents:
-			s.handleChainBlock(b)
-		default:
+		// Always process block event if there is any, we can add one above or external
+		// services can add several blocks during message processing.
+	syncLoop:
+		for {
+			select {
+			case b := <-s.blockEvents:
+				s.handleChainBlock(b)
```

Comment:
Shouldn't we only do this once per block batch?

Reply:
Why? And what should we do with notification from the block that was accepted by consensus otherwise?

Reply:
It's handled as usual. I mean you can get a number of events in this loop, but you only care about the latest one because other initializations don't make much sense, they're obsolete by the time you get to the next block event (really soon in this loop if we have a number of blocks in the queue).

At the same time, `blockEvents` only has a len of 1. And after the rework we no longer push a block in the same context, so maybe it's not an issue.

Reply:
Got your idea, but yes,

> blockEvents only has a len of 1

we don't have any buffer, so blockchain can be only two blocks further than the current one that we've got from `case b := <-s.blockEvents:`, so I'm not sure whether it's worth to save them and pick up only latest.

Reply:
And the block processing time is still larger than `handleChainBlock` takes.

Reply:
Although let's implement this improvement, it won't hurt.

Reply:
Fixed in a separate commit.